dbms/src/Functions/FunctionBase64Conversion.h

```diff
@@ -0,0 +1,150 @@
+    {
+        return true;
+    }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        size_t number_of_arguments = arguments.size();
+
+        if (number_of_arguments != 1)
+            throw Exception(
+                "Function " + getName() + " needs exactly one argument of type String", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        if (!WhichDataType(arguments[0].type).isString())
+            throw Exception(
+                "Illegal type " + arguments[0].type->getName() + " of 1 argument of function " + getName() + ". Must be String.",
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t input_rows_count) override
+    {
+        const ColumnPtr column_string = block.getByPosition(arguments[0]).column;
+        const ColumnString * input = checkAndGetColumn<ColumnString>(column_string.get());
+
+        if (!input)
+            throw Exception(
+                "Illegal column " + block.getByPosition(arguments[0]).column->getName() + " of first argument of function " + getName(),
+                ErrorCodes::ILLEGAL_COLUMN);
+
+        auto dst_column = ColumnString::create();
+        auto & dst_data = dst_column->getChars();
+        auto & dst_offsets = dst_column->getOffsets();
+
+        size_t reserve = ceil(input->getChars().size() * Func::buffer_size_multiplier + input->size());
+        dst_data.resize(reserve);
+        dst_offsets.resize(input_rows_count);
+
+        const ColumnString::Offsets & src_offsets = input->getOffsets();
+
+        auto source = reinterpret_cast<const char *>(input->getChars().data());
+        auto dst = reinterpret_cast<char *>(dst_data.data());
+        auto dst_pos = dst;
+
+        size_t src_offset_prev = 0;
+
+        int codec = getCodec();
+        for (size_t row = 0; row < input_rows_count; ++row)
+        {
+            size_t srclen = src_offsets[row] - src_offset_prev - 1;
+            size_t outlen = 0;
+
+            if constexpr (std::is_same_v<Func, Base64Encode>)
+            {
+                base64_encode(source, srclen, dst_pos, &outlen, codec);
+            }
+            else
+            {
+                if (!base64_decode(source, srclen, dst_pos, &outlen, codec))
+                {
+                    throw Exception("Failed to " + getName() + " input '" + String(source, srclen) + "'", ErrorCodes::INCORRECT_DATA);
+                }
+            }
+
+            source += srclen + 1;
+            dst_pos += outlen + 1;
+
+            dst_offsets[row] = dst_pos - dst;
+            src_offset_prev = src_offsets[row];
+        }
+
+        dst_data.resize(dst_pos - dst);
+
+        block.getByPosition(result).column = std::move(dst_column);
+    }
+
+private:
+    static int getCodec()
+    {
+#if __SSE4_2__
```

Comment:
Why don't use runtime CPU detection?
If I understand correctly, you should compile base64 library with all supported instructions (AVX2 on x86_64, NEON or arm).

Reply:
In order to do it, we need to add compilation flags for AVX/NEON/SSSE3 features. When I was adopting this library to CH, I've only reused existing SSE4.1 and SSE4.2 detection features. 
Should we add detection for other as well?

Reply:
it looks like you can enable all of them at compile time, so preprocessor will not throw away corresponding code blocks. If you will not set any FORCE flags it will do some autodetection at runtime. Check here: 
https://github.com/aklomp/base64#flags

>Set flags to 0 for the default behavior, which is runtime feature detection on x86, a compile-time fixed codec on ARM, and the plain codec on other platforms.